### ISSUE TITLE
Self-erasing functions not transformed correctly

### ISSUE THREAD

**Issue by wychen** [facebook/prepack#570]
Input code:
```
(function() {
  function init() {
    init = function() {};
    v = {};
  }
  function work() {
    a = (init(), v);
    b = (init(), v);
    c = a == b;
  }
  global.work = work;
})();
```

Note that `init()` is a self-erasing function, so that `v` is initialized only once. When running `work()`, `c` should be `true`.

The transformed code is:
```
(function () {
  function _0() {
    a = (_1(), v);
    b = (_1(), v);
    c = a == b;
  }

  function _1() {
    $0 = function () {};

    v = {};
  }

  var $0 = _1;
  work = _0;
})();
```

Note that function `_1()` is no longer a self-erasing function, so `v` is initialized twice, making `a` and `b` different objects. When running `work()`, `c` becomes `false`, which is different from the original code.

**Comment by wychen** [comment]
This self-erasing pattern is used a lot in the JS code generated by [GWT](http://www.gwtproject.org/). This bug makes prepack unusable to reprocess the output of GWT.

**Comment by nicolo-ribaudo** [comment]
Hi, would a pull request for this bug be accepted?

**Comment by wychen** [comment]
@nicolo-ribaudo I think it's always a step forward to send it out and get some peer feedback.

**Comment by cblappert** [comment]
We would definitely welcome someone to come and fix this bug! If you would like you could also create an empty pull request or comment on the issue with your idea for fixing the problem so that you can get feedback on it before coding up the fix!

**Comment by nicolo-ribaudo** [comment]
This is what I think could solve this issue:

I noticed that inside `Serializer#_spliceFunctions()` prepack generates a new identifier for every modified value. (Line 1254)
Non-primitive values (e.g. objects and functions) are serialized by reference (identifier) instead of value, so it gets generated a new identifier whose value is a reference to the old one.
This is noticeable when compiling this code:
```js
(function() {
  var obj = {
    fn() {
      obj = { fn() {} };
    }
  };

  inspect = () => obj.fn();
})();
```
```js
(function () {
  function _0() {
    return $0.fn();
  }

  function _2() {
    $0 = {
      fn() {}

    };
  }

  var _1 = {
    fn: _2
  };
  var $0 = _1;
  inspect = _0;
})();
```

Do you see that the object is defined as `_1` and then aliased to `$0`? This is not a problem for objects, but it yelds problems when it happens for a function (because the function call references the function's original name).

The solution I propose is to not alias these values (e.g. use `_1` instead of `$0`), so that the issue doesn't happens. What do you think about this idea?

**Comment by NTillmann** [comment]
The generation of new identifiers inside Serializer#_spliceFunctions() is needed to distinguish different captured activation scopes. That by itself isn't wrong. What was wrong is that for some reason not all occurrences of `init` got the same treatment --- after all, they should have all resolved to the same scope.

I looked into it a bit more and figure out that it was due to the ordering in which the functions were explored during the visiting phase, which resulted in the creation of different binding instances, and only the last one won. I fix this in #676 and add a regression test.
